### PR TITLE
added plot_facts service in database_manager

### PR DESCRIPTION
**Pull request by shipra25jain**

### CODE REVIEW

File: database_manager/src/run_server.cpp [hunk 1]
```diff
@@ -19,6 +20,10 @@
 #include "toaster_msgs/Fact.h"
 #include "toaster_msgs/Id.h"
 #include "toaster_msgs/FactList.h"
+#include <fstream>
```
Comment (Greg8978) [inline database_manager/src/run_server.cpp:22]: Please chack that those 3 includes are needed. They are probably already included in other included files.

**Comment by sdevin** [comment]
Several comments:
- PlotFactsDB: you should use more explicit names than string a and string b!
- function plot_facts_db: some comments in the code can be welcomed
- in the database we try to have less services as possible in order to ensure the speed of the node: should we integrate this service to another (the execute service)?

File: toaster_msgs/srv/PlotFactsDB.srv [hunk 2]
```diff
@@ -0,0 +1,7 @@
+string subjectID
+string targetID
+string a
```
Comment (Greg8978) [inline toaster_msgs/srv/PlotFactsDB.srv]: please put a more specific name than a and b.
Plus if it's time you shouldn't use string...

File: database_manager/src/run_server.cpp [hunk 3]
```diff
@@ -1854,6 +1854,97 @@ bool execute_db(toaster_msgs::ExecuteDB::Request &req, toaster_msgs::ExecuteDB::
    return true;
 }
 
+////////////////////////////////////////////////
+//for graphical representation of facts/////////
+////////////////////////////////////////////////
+
+//for graphical representation
+
+bool plot_facts_db(toaster_msgs::PlotFactsDB::Request &req, toaster_msgs::PlotFactsDB::Response &res){
```
Comment (Greg8978) [inline database_manager/src/run_server.cpp:1862]: Please add some more comments and space to make this code easy to understand

File: toaster_msgs/srv/PlotFactsDB.srv (hunk 2 above)
Comment (shipra25jain) [inline toaster_msgs/srv/PlotFactsDB.srv]: yeah, I am changing it to timeStart and timeEnd. And type as uint32.

File: database_manager/src/run_server.cpp [hunk 4]
```diff
@@ -1854,6 +1854,97 @@ bool execute_db(toaster_msgs::ExecuteDB::Request &req, toaster_msgs::ExecuteDB::
...
+
+     std::string sql;
+     char **pazResult;
+      int pnRow;
+      int pnColumn ;
+      char *err_msg = NULL;
+      int ret;
+      long long int i;
```
Comment (Greg8978) [inline database_manager/src/run_server.cpp:1870]: I don't think you need a long long here.
An unsigned int should be enough?

Comment (shipra25jain) [inline database_manager/src/run_server.cpp:1870]: with int and long, it did not work so i used long long.

Comment (Greg8978) [inline database_manager/src/run_server.cpp:1870]: Ok, ignore my comment then :)

Comment (shipra25jain) [inline database_manager/src/run_server.cpp:1870]: okay, I will check once.

**Comment by sdevin** [comment]
Concerning the commit to correct time issues:
- why is the modif of it to itt necessary?
- should we use ros:time::now instead of curTimeOfDay? (as it is done in other parts of toaster)

**Comment by shipra25jain** [comment]
- It is a nested for loop with same "it" iterator used for parent loop
- But the curTimeOfDay has been used here :
  https://github.com/Greg8978/toaster/blob/master/pdg/src/HumanReader.cpp

**Comment by sdevin** [comment]
- yes it is true ignore my first comment
- this is the code for the isPresent computation, however, the time put into entities and so used to fill facts is ros::time::now() (e.g. https://github.com/Greg8978/toaster/blob/master/pdg/src/Pr2RobotReader.cpp l.89)

File: database_manager/src/run_server.cpp [hunk 5]
```diff
@@ -1918,19 +1913,14 @@ bool plot_facts_db(toaster_msgs::PlotFactsDB::Request &req, toaster_msgs::PlotFa
...
-       std::ofstream myfile;
-       myfile.open ("fact.dat"); // you can mention the full complete path of file like "/home/jshipra/fact.dat"
-       long long int diff = b - a ;
-       for(i = 0 ; i< diff ; i=i+100000)
-      {  long long int x = a + i;
-		  int k = 0;
-        for(i = 0 ; i< diff ; i=i+100000)
-      {  long long int x = a + i;
+      std::ofstream myfile;
+      myfile.open ("/home/jshipra/fact.dat");
```
Comment (Greg8978) [inline database_manager/src/run_server.cpp]: use relative path.
If I run this code on my computer it won't know what is /home/jshipra.
Maybe use something like: ros:: package::getPath("database_manager") << "/plot/fact << req.reqFact << .dat";

**Comment by Greg8978** [comment]
Ok, it's good for me.
@sdevin if it's fine for you I will accept the pull request :)

File: database_manager/src/run_server.cpp [hunk 6]
```diff
@@ -775,14 +776,17 @@ bool remove_facts_to_agent_db(std::string agentId, std::vector<toaster_msgs::Fac
...
-                       + boost::lexical_cast<std::string>(it->subjectId) + "','!"
-                       + (std::string)it->property + "','"
-                       + (std::string)it->propertyType + "','"
-                       + boost::lexical_cast<std::string>(it->targetId) + "',"
-                       + boost::lexical_cast<std::string>(it->factObservability) + ","
-                       + boost::lexical_cast<std::string>(it->confidence) + ","
-                       + boost::lexical_cast<std::string>(it->time) + ")";
+            timeval curTime;
+            gettimeofday(&curTime, NULL);
+            unsigned long now = curTime.tv_sec * pow(10,9) + curTime.tv_usec;
```
Comment (Greg8978) [inline database_manager/src/run_server.cpp]: Please make this last change:
facts is ros::time::now() (e.g. https://github.com/Greg8978/toaster/blob/master/pdg/src/Pr2RobotReader.cpp l.89)

**Comment by Greg8978** [comment]
Ok, almost there!
Squash some of your commits that are not relevant like:
"Update run_server.cpp"

And please, don't forget to do a pull --rebase to keep the commit history tree straigh.

Thanks :)

**Pull request closed without merging**